### ISSUE TITLE
call API observe more than once will lead to crash

### ISSUE THREAD

**Issue by learscnuli** [etcd-cpp-apiv3/etcd-cpp-apiv3#224]
etcd api: etcd-cpp-apiv3-v0.14.2
grpc: 1.50.1

if we call etcd->observe("xxxxx") twice, we will enter the function: void etcdv3::AsyncObserveAction::waitForResponse().
In this function, we will get the function call routine: 

```
---------------------------------
waitForResponse in:
response_reader->Read
cq_.Next
response_reader->Read
waitForResponse out:

waitForResponse in:
response_reader->Read
cq_.Next
response_reader->Read
waitForResponse out:
---------------------------------
```

there's a "Read->Read->Next".
![企业微信截图_16855879466956](https://github.com/etcd-cpp-apiv3/etcd-cpp-apiv3/assets/13833278/bfd8b33a-bd8b-4887-82d3-52d25aa42526)

and crash here: 
![企业微信截图_16855879921386](https://github.com/etcd-cpp-apiv3/etcd-cpp-apiv3/assets/13833278/d352cfc2-d713-40f9-a4cd-f44d70db632f)

plz fix the code as below:

```
void etcdv3::AsyncObserveAction::waitForResponse()
{
  void* got_tag;
  bool ok = false;

  if (isCancelled.load()) {
    status = grpc::Status::CANCELLED;
  }
  if (!status.ok()) {
    return;
  }

  response_reader->Read(&reply, (void *)this);
  if (cq_.Next(&got_tag, &ok) && ok && got_tag == (void*)this) {
    auto response = ParseResponse();
    if (response.get_error_code() == 0) {
      // issue the next read
      // response_reader->Read(&reply, (void *)this);            // please comment this line
    } else {
      this->CancelObserve();
    }
  } else {
    this->CancelObserve();
    status = grpc::Status::CANCELLED;
  }
}
```

thank u.

**Comment by sighingnow** [comment]
Hi @learscnuli,

Is there any code snippets that can help to reproduce the issue? "we call etcd->observe("xxxxx") twice" is not very clear to me. Could you please provide me more context and it it is possible, a minimal example code that can directly reproduce the issue would be great.

Thanks!

**Comment by learscnuli** [comment]
we use these functions to elect a service leader.
herer is a simple example run in UE when I repredoce the bug.

1. each instance will obtain a leaseid
		auto Response = EtcdClient->leasegrant(TTL).get();
		int64 LeaseID = Response.value().lease();

2. keepalive the leaseid and save the keepalive
	auto Handler = [this, LeaseID](std::exception_ptr exception)
	{
		KeepAliveExptHandler(LeaseID, exception);
	};

	TSharedPtr<etcd::KeepAlive> KeepAlive(new etcd::KeepAlive(*EtcdClient, Handler, TTL, LeaseID));
	LeaseID2KeepAliveMap.Add(LeaseID, KeepAlive);

3. register an instance
			Rsp = EtcdClient->set("/xx/InstanceID1",
				"Instance_value1",
				LeaseID);

4. campain the Instance as "MyService"
std::string ServiceName = "MyService";
auto Response = EtcdClient->campaign(ServiceName, LeaseID, "InstanceID1").get();

5. observe the leader change

			etcd::Client ThreadEtcdClient("http://xx.xx.xx.xx:2379");
			std::unique_ptr<etcd::SyncClient::Observer> Observer = ThreadEtcdClient.observe(ServiceName);
			check(Observer != nullptr);

			etcd::Response WResponse = Observer->WaitOnce();
			FString NewLeader(WResponse.value().as_string().c_str());
			// OnServiceLeaderChanged(NewLeader);
                        // some log here


			etcd::Response R2 = Observer->WaitOnce();  // will crash here
			FString NewLeader2(WResponse.value().as_string().c_str());
			OnServiceLeaderChanged(NewLeader2);
			
I notice that we don't have a testcase of observe() function, could you plz add a new one as well?
thks a lot.

**Comment by sighingnow** [comment]
Will be addressed by #231.
